Bugzilla/API/V1/Github.pm

```diff
@@ -164,7 +171,167 @@ sub pull_request {
+
+  # Return early if not a push or ping event
+  my $event = $self->req->headers->header('X-GitHub-Event');
+  if (!$event || ($event ne 'push' && $event ne 'ping')) {
+    return $self->code_error('github_push_comment_not_push');
+  }
+
+  # Verify that signature is correct based on shared secret
+  if (!$self->_verify_signature) {
+    return $self->code_error('github_push_comment_mismatch_signatures');
+  }
+
+  # If event is a ping and we passed the signature check
+  # then return success
+  if ($event eq 'ping') {
+    return $self->render(json => {error => 0});
+  }
+
+  # Parse push commit title for bug ID
+  my $payload = $self->req->json;
+  if ( !$payload
+    || !$payload->{ref}
+    || !$payload->{repository}
+    || !$payload->{repository}->{full_name}
+    || !$payload->{pusher}
+    || !$payload->{pusher}->{name}
+    || !$payload->{commits})
+  {
+    return $self->code_error('github_push_comment_invalid_json');
+  }
+
+  my $ref        = $payload->{ref};
+  my $repository = $payload->{repository}->{full_name};
+  my $pusher     = $payload->{pusher}->{name};
+  my $commits    = $payload->{commits};
+
+  # Return success early if there are no commits or the ref is not a branch
+  if (!@{$commits} || $ref !~ /refs\/heads\//) {
+    return $self->render(json => {error => 0});
+  }
+
+  # Keep a list of bug ids that need to have comments added.
+  # We also use this for sending email later.
+  # Use a hash so we don't have duplicates. If multiple commits
+  # reference the same bug ID, then only one comment will be added
+  # with the text combined.
+  # When the comment is created, we will store the comment id to
+  # return to the caller.
+  my %update_bugs;
+
+  # Create a separate comment for each commit
+  foreach my $commit (@{$commits}) {
+    my $message = $commit->{message};
+    my $url     = $commit->{url};
+
+    if (!$url || !$message) {
+      return $self->code_error('github_pr_invalid_json');
+    }
+
+    # Find bug ID in the title and see if bug exists
+    my ($bug_id) = $message =~ /\b[Bb]ug[ -](\d+)\b/;
+    next if !$bug_id;
+
+    my $comment_text = "Authored by https://github.com/$pusher\n$url\n$message";
+
+    $update_bugs{$bug_id} ||= [];
+    push @{$update_bugs{$bug_id}}, {text => $comment_text};
+  }
+
+  # If no bugs were found, then we return an error
+  if (!keys %update_bugs) {
+    return $self->code_error('github_push_comment_bug_not_found');
+  }
+
+  # Set current user to automation so we can add comments to private bugs
+  my $auto_user = Bugzilla::User->check({name => 'automation@bmo.tld'});
+  $auto_user->{groups}       = [Bugzilla::Group->get_all];
+  $auto_user->{bless_groups} = [Bugzilla::Group->get_all];
+  Bugzilla->set_user($auto_user);
+
+  my $dbh = Bugzilla->dbh;
+  $dbh->bz_start_transaction;
+
+  # Actually create the comments in this loop
+  foreach my $bug_id (keys %update_bugs) {
+    my $bug = Bugzilla::Bug->new({id => $bug_id, cache => 1});
+    next if $bug->{error};
```

Comment:
What does this `next if` catch? Invalid bug number?

Reply:
If the bug is not visible or does not exist, etc. then $bug->{error} will be populated. We then pass on to the next commit silently and not cause an exception. If we throw an exception if will not get to the other commits which may have valid IDs. We just treat the invalid IDs as if they commit did not match our bug id format rules.